client/cli/src/lib.rs

```diff
@@ -177,461 +179,393 @@ fn is_node_name_valid(_name: &str) -> Result<(), &str> {
 		.about(version.description)
-		.version(&(full_version + "\n")[..])
-		.setting(AppSettings::GlobalVersion)
-		.setting(AppSettings::ArgsNegateSubcommands)
-		.setting(AppSettings::SubcommandsNegateReqs)
-		.get_matches_from(args);
-	let cli_args = CoreParams::<CC, RP>::from_clap(&matches);
-	fdlimit::raise_fd_limit();
+		.version(full_version.as_str());
 
-	let args = match cli_args {
-		params::CoreParams::Run(params) => ParseAndPrepare::Run(
-			ParseAndPrepareRun { params, impl_name, version }
-		),
-		params::CoreParams::BuildSpec(params) => ParseAndPrepare::BuildSpec(
-			ParseAndPrepareBuildSpec { params, version }
-		),
-		params::CoreParams::ExportBlocks(params) => ParseAndPrepare::ExportBlocks(
-			ParseAndPrepareExport { params, version }
-		),
-		params::CoreParams::ImportBlocks(params) => ParseAndPrepare::ImportBlocks(
-			ParseAndPrepareImport { params, version }
-		),
-		params::CoreParams::CheckBlock(params) => ParseAndPrepare::CheckBlock(
-			CheckBlock { params, version }
-		),
-		params::CoreParams::PurgeChain(params) => ParseAndPrepare::PurgeChain(
-			ParseAndPreparePurge { params, version }
-		),
-		params::CoreParams::Revert(params) => ParseAndPrepare::RevertChain(
-			ParseAndPrepareRevert { params, version }
-		),
-		params::CoreParams::Custom(params) => ParseAndPrepare::CustomCommand(params),
-	};
-	init_logger(args.shared_params().and_then(|p| p.log.as_ref()).map(|v| v.as_ref()).unwrap_or(""));
-	args
+	T::from_clap(&app.get_matches_from(iter))
 }
 
-/// Returns a string displaying the node role, special casing the sentry mode
-/// (returning `SENTRY`), since the node technically has an `AUTHORITY` role but
-/// doesn't participate.
-pub fn display_role<A, B, C>(config: &Configuration<A, B, C>) -> String {
-	if config.sentry_mode {
-		"SENTRY".to_string()
-	} else {
-		format!("{:?}", config.roles)
-	}
-}
-
-/// Output of calling `parse_and_prepare`.
-#[must_use]
-pub enum ParseAndPrepare<'a, CC, RP> {
-	/// Command ready to run the main client.
-	Run(ParseAndPrepareRun<'a, RP>),
-	/// Command ready to build chain specs.
-	BuildSpec(ParseAndPrepareBuildSpec<'a>),
-	/// Command ready to export the chain.
-	ExportBlocks(ParseAndPrepareExport<'a>),
-	/// Command ready to import the chain.
-	ImportBlocks(ParseAndPrepareImport<'a>),
-	/// Command to check a block.
-	CheckBlock(CheckBlock<'a>),
-	/// Command ready to purge the chain.
-	PurgeChain(ParseAndPreparePurge<'a>),
-	/// Command ready to revert the chain.
-	RevertChain(ParseAndPrepareRevert<'a>),
-	/// An additional custom command passed to `parse_and_prepare`.
-	CustomCommand(CC),
-}
+pub fn run<F, G, E, F2, F3, F4, T1, T2, T3, T5>(
+	core_params: CoreParams,
+	new_light: F2,
+	new_full: F3,
+	spec_factory: F,
+	builder: F4,
+	impl_name: &'static str,
+	version: &VersionInfo,
+) -> error::Result<()>
+where
+	F: FnOnce(&str) -> Result<Option<ChainSpec<G, E>>, String>,
+	F2: FnOnce(Configuration<G, E>) -> Result<T1, sc_service::error::Error>,
```

Comment:
I'm really not a fan of this callback style where we're hard-coding the fact that `new_full` and `new_light` take a configuration and return a service.

The reason for not using callbacks originally is so that users can straight up do a function call, which to me is much more intuitive than having to pass a closure that is then called by something else.

When later we want to pass another parameter to `new_full`, what are we going to do? Pass this parameter to `run` so that it is then passed to the callback?

Yes this PR reduces code duplication, but IMO it's not a good thing here.

Reply:
But... but... ☝️  It was already a closure/callback before. I didn't change that ❓ what do you mean❓  https://github.com/paritytech/substrate/blob/master/bin/node/cli/src/cli.rs#L101

Reply:
Oh, just because it's in master doesn't mean I like it :sweat_smile: 
Of course that's just my opinion, but as the one who has done the previous refactoring here, I was trying to remove these callbacks-like structure. The one you linked to was hard to remove at the time, for a reason I can't recall.

Reply:
haha :grin: ok! Actually I just tried to remove the callback and use a trait instead (so you can override functions in the trait) but it's actually very complicated because of the return types: `new_full` and `new_start` return 2 different things in bin/node: one is Service and the other is impl AbstractService...

I couldn't find a solution to that but in any case: I'm trying also to reduce the complexity of the signatures.

Reply:
> If later we want to pass another parameter to new_full, what are we going to do? Pass this parameter to run so that it is then passed to the callback?

https://github.com/paritytech/substrate/pull/4692#discussion_r368922595